Fix Kubeproxy to work for Windows Kernel mode

madhanrm opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
-->

**What this PR does / why we need it**:
Kubeproxy doenst work for with windows kernel mode. This PR adds a Kernel Proxy for windows to use the underlying platform features.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
#49666
**Special notes for your reviewer**:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
```

madhanrm committed:
Add Windows Kernel Proxy support

	Windows Kernel now exposes "Internal Load Balancing"
	using VFP (Virtual Filtering Platform) part of Virtual Switch. An inbuild
	windows service HNS (Host Networking Service) acts as interface to program
	the VFP. VFP is synonymous to iptables in functionality. HNS uses json based
	data as input.

	With the help of the interface available in github.com/Microsoft/hcsshim,
	these APIs are exposed to the world in github to program HNS and use
	the feature.

	*** More info about the changes in this PR ***
	(1) For every endpoint available in the system, an HNS Endpoint is added
	    (1.a) for local endpoints, a local HNS Endpoint would already exist, as part of
            container creation.
	    (1.b) For all remote endpoints, a remote HNS Endpoint is created via HNS

	(2) For every Service, a HNS ILB LoadBalancer is added referring the endpoints
	    created in (1)
		Sample Input to HNS:
		{
 	       "Policies":  [
        	                 {
                	             "ExternalPort":  80,
                        	     "InternalPort":  80,
	                             "Protocol":  6,
        	                     "Type":  "ELB",
                	             "VIPs":  [
                        	                  "11.0.98.129"
                                	      ]
	                         }
        	             ],
	        "References":  [
                           "/endpoints/ca8b877b-ab90-499a-bc0e-7d736c425632",
                           "/endpoints/ee0ef08b-8434-4f8b-b748-393884e77465"
        		]
    		}

	(2-a) This is done for Cluster IP, LoadBalancer Ingress IP, NodePort, External IP

	Following the regular service and endpoint updates,
	the HNS is notified of the updates and the system is kept in sync.

madhanrm commented:
Do not merge yet.

E2e testing is not complete

freehan commented:
/ok-to-test

dineshgovindasamy commented:
Tagging @jdumars  for FYI

jdumars commented:
/sig node
/sig windows

jdumars commented:
/cc @mikedanese

madhanrm commented:
Pending Vendor commit

JMesser81 commented:
@freehan - let's discuss at next sig-network meeting. We believe this is a low risk change which does not affect Linux at all or current Windows implementations using OVN or Apprenda solution.

madhanrm committed:
Fix Bazel build

aserdean commented:
(this is not a technical review)

Please break: https://github.com/kubernetes/kubernetes/pull/51064/commits/6e6d1323fb84980677cc7c79faf62fa68850ada9 into
smaller commits, as described in https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#6-squashing-and-commit-titles.
It makes it easier for other people to review it and understand your logic
behind the changes.

Please add relevant commit messages to each individual commit, it helps
other people understand how you are trying to achieve your goal and of
course keep track of the evolution of the code.

madhanrm committed:
Vendor changes

Vendoring (NEW) in github.com/Microsoft/hcsshim

jdumars commented:
/cc @thockin

madhanrm commented:
/test pull-kubernetes-cross
/test pull-kubernetes-kubemark-e2e-gce
/test pull-kubernetes-e2e-gce-gpu
/test pull-kubernetes-bazel
/test pull-kubernetes-verify

madhanrm commented:
/test pull-kubernetes-federation-e2e-gce

madhanrm committed:
Add exception to golint check

(*) Fix cleanup of NodePort resources. (*) Fix the logic to select existing policies

Fix review comment

Fix Bazel

Update GoDep License

Fix NodePort forwarding to target port

Fix Darwin Build break. +1

Implement IsCompatible to validate kernel support for kernel mode

madhanrm commented:
/test pull-kubernetes-federation-e2e-gce

thockin commented on cmd/kube-proxy/app/server.go:
```diff
@@ -79,8 +72,9 @@ import (
 const (
-	proxyModeUserspace = "userspace"
-	proxyModeIPTables  = "iptables"
+	proxyModeUserspace   = "userspace"
```
I don't like how non-obvious it is that these don't all work on any give platform.  Can we move these constants into the per-OS files and call a function from there to get a list of options to show in strings?  We should also deal with flags and componentconfig that do not map across OSes.

How much code reuse are we really getting by having these things live in the same sub-repo?  Would it be better to just make a win-kube-proxy ?  So much of what is left here is still linux-specific

madhanrm replied:
I would prefer a plugin model here, to provide interfaces for Service/Endpoint Updates, and just customize it in the plugin, rather than bringing a lot of code in to Kubernetes, along with the vendor changes.

thockin reviewed (commented):
Overall this is fine - I can't realistically review the Windows side of it, anyway.  I think it doesn't go far enough, though.  Can we split the OS-specifics even more?

thockin commented:
I'll LGTM for the sake of code freeze, but not approve until we decide on remaining OS-specific stuff

thockin commented:
/lgtm

dcbw commented on pkg/proxy/winkernel/proxier.go:
```diff
@@ -0,0 +1,1105 @@
+// CanUseWinKernelProxier returns true if we should use the Kernel Proxier
+// instead of the "classic" userspace Proxier.  This is determined by checking
+// the windows kernel version and for the existence of kernel features.
+func CanUseWinKernelProxier(kcompat KernelCompatTester) (bool, error) {
```
All this WindowsKernelCompatTester stuff doesn't actually do anything that I can see; it always returns nil for IsCompatible().  Perhaps just remove it?

madhanrm replied:
There is a TODO to add this function, since old windows kernel wouldn't support this feature and hence kernel proxy wouldn't work for older versions

dcbw replied:
@madhanrm I'm of the opinion that you should remove the bits that don't do anything for now, and add them back with a new PR when you have implemented them fully.  It keeps the original PR smaller and easier for review, and prevents forgotten bits of code from getting committed that never end up getting used.

madhanrm replied:
I have taken care of it, and finished the implementation of this NO-OP method.

dcbw commented on pkg/proxy/winkernel/proxier.go:
```diff
@@ -0,0 +1,1105 @@
+	ip         string
+	port       uint16
+	isLocal    bool
+	macAddress string
```
macAddress does not appear to be used anywhere.  If it's always going to be constant, just put that constant into the HNSEndpoint{} creation instead of carrying it around in the endpointsInfo.

madhanrm replied:
It may appear like that for now, but I would like to be this way, so that, when we have info about the macAddress, this can be populated with right data.

dcbw commented on pkg/proxy/winkernel/proxier.go:
```diff
@@ -0,0 +1,1105 @@
+}
+
+type updateEndpointMapResult struct {
+	hcEndpoints       map[types.NamespacedName]int
```
It doesn't look like hcEndpoints is used anywhere in this proxier.  Perhaps just remove the healthcheck stuff if Windows doesn't use it?

madhanrm replied:
I actually missed out the code. Will add it back to use healthcheck

dcbw commented on pkg/proxy/winkernel/proxier.go:
```diff
@@ -0,0 +1,1105 @@
+	serviceUpdateResult := proxier.updateServiceMap()
+	endpointUpdateResult := proxier.updateEndpointsMap()
+
+	staleServices := serviceUpdateResult.staleServices
```
it doesn't look like staleServices gets used at all?

madhanrm replied:
I'll have to look into this again, to see if I missed anything with this staleServices

dcbw commented on pkg/proxy/winkernel/proxier.go:
```diff
@@ -0,0 +1,1105 @@
+
+	// Program HNS by adding corresponding policies for each service.
+	for svcName, svcInfo := range proxier.serviceMap {
+		if svcInfo.policyApplied {
```
This looks kinda odd.  In what cases would we have effectively the same svcInfo object under two different svcName keys in the serviceMap?

madhanrm replied:
I'll look into this.

madhanrm replied:
@dcbw Where do you see that assumption?

madhanrm commented on hack/.golint_failures:
```diff
@@ -301,6 +301,7 @@ pkg/proxy
 pkg/proxy/iptables
 pkg/proxy/userspace
 pkg/proxy/util
+pkg/proxy/winkernel
```
This happens for other projects as well. I just added the exception just like others

dcbw replied:
@madhanrm did you add it because the lint failed, or just because others had?

madhanrm replied:
lint failed, as it did failed on other projects. So added it to the exception for now.

cmluciano replied:
I ran this locally without the exception and a lot of the errors are coming from the general proxier pkg not have comments and whatnot. We probably should have a TODO to clean these up later.

madhanrm commented on cmd/kube-proxy/app/server.go:
```diff
@@ -416,191 +410,6 @@ func createClients(config componentconfig.ClientConnectionConfiguration, masterO
 }
 
-// NewProxyServer returns a new ProxyServer.
-func NewProxyServer(config *componentconfig.KubeProxyConfiguration, cleanupAndExit bool, scheme *runtime.Scheme, master string) (*ProxyServer, error) {
```
Its the first set of lines which are generic. getProxymode is also local, and it made sense to move this entire method.

dcbw replied:
@madhanrm yes, the first set is generic and could be kept generic between the implementations.  These lines:

```
	if config == nil {
		return nil, errors.New("config is required")
	}

	if c, err := configz.New("componentconfig"); err == nil {
		c.Set(config)
	} else {
		return nil, fmt.Errorf("unable to register configz: %s", err)
	}
<snip>
	client, eventClient, err := createClients(config.ClientConnection, master)
	if err != nil {
		return nil, err
	}

	// Create event recorder
	hostname := utilnode.GetHostname(config.HostnameOverride)
	eventBroadcaster := record.NewBroadcaster()
	recorder := eventBroadcaster.NewRecorder(scheme, v1.EventSource{Component: "kube-proxy", Host: hostname})

	nodeRef := &v1.ObjectReference{
		Kind:      "Node",
		Name:      hostname,
		UID:       types.UID(hostname),
		Namespace: "",
	}

	var healthzServer *healthcheck.HealthzServer
	var healthzUpdater healthcheck.HealthzUpdater
	if len(config.HealthzBindAddress) > 0 {
		healthzServer = healthcheck.NewDefaultHealthzServer(config.HealthzBindAddress, 2*config.IPTables.SyncPeriod.Duration, recorder, nodeRef)
		healthzUpdater = healthzServer
	}
```

Are all the same, and that's half of the function.  I think it makes sense to keep that generic.

madhanrm replied:
When we have another abstract layer, to abstract out the common parts, all this would go away. So it doesn't make sense to make this change now.

cmluciano replied:
This is breaking the cross-build as well

```
I0907 19:34:13.959] cmd/kube-proxy/app/server.go:224: undefined: NewProxyServer
```

madhanrm replied:
Is there a way to fix the cross build for Darwin. I have added a build tag for server_linux.go to // +build !windows. I expect this to work with all OS except windows. Why doenst it work for Darwin? Any clues?

madhanrm replied:
Took care of it by renaming the files. Weird though.

madhanrm commented on cmd/kube-proxy/app/server.go:
```diff
@@ -664,7 +473,7 @@ func (s *ProxyServer) Run() error {
 
 	// Tune conntrack, if requested
-	if s.Conntracker != nil && goruntime.GOOS != "windows" {
+	if s.Conntracker != nil {
```
Yes, to keep the movement of the code for this, I have kept this assumption. Will add a comment in the code as well.

feiskyer commented on cmd/kube-proxy/app/server_linux.go:
```diff
@@ -0,0 +1,298 @@
+	"github.com/golang/glog"
+)
+
+func NewProxyServer(config *componentconfig.KubeProxyConfiguration, cleanupAndExit bool, scheme *runtime.Scheme, master string) (*ProxyServer, error) {
```
nit: missing interface docs

feiskyer commented on cmd/kube-proxy/app/server_windows.go:
```diff
@@ -0,0 +1,187 @@
+	var serviceEventHandler proxyconfig.ServiceHandler
+	var endpointsEventHandler proxyconfig.EndpointsHandler
+
+	proxyMode := getProxyMode(string(config.Mode), winkernel.WindowsKernelCompatTester{})
```
throw out an error if a wrong mode is configured? e.g. iptables is not supported

madhanrm replied:
If incorrect mode is passed, it would fall back to userspace

feiskyer reviewed (commented):
There are much code repetition across different proxy modes, e.g. the services and endpoints watchers are same. I think it's better to make another abstract layer and let each mode only focus on core logic. 

This is not specific to this PR, but it could help of extending new proxy modes.

thockin commented:
This has O(hours) to get approved.  Ping me on slack or something if you get it patched up and ready.

jdumars commented:
@madhanrm do you have any additional commits to make here?  As Tim said, we're in the final hours before code freeze.

madhanrm commented:
@jdumars  & @thockin, am done with my changes. Do you want me to squeeze the changes? I had rebased yesterday with the latest changes

madhanrm commented:
/test pull-kubernetes-unit

madhanrm commented:
Commit 19220bb0333a850a7adf1f8d10181731900032d3 fixes #51680  to unblock my test

thockin commented:
/lgtm

jdumars commented:
/retest

calebamiles commented:
removing the `do-no-merge` label added due to missing release note label

madhanrm commented:
/test pull-kubernetes-e2e-kops-aws

mikedanese commented:
Please squash fixup commits

madhanrm commented:
/retest

madhanrm commented:
@mikedanese  squashed up fixed up commits

madhanrm commented:
/test pull-kubernetes-e2e-kops-aws

madhanrm commented:
/test pull-kubernetes-e2e-kops-aws

madhanrm commented:
Rebased again and fixed up the godep license.

madhanrm commented:
/test pull-kubernetes-e2e-kops-aws

madhanrm commented:
@thockin  can you re-add lgtm label? It got removed again, after a rebase.

dims commented:
@madhanrm "make cross" is failing, need that fixed first i think

dcbw commented:
> There are much code repetition across different proxy modes, e.g. the services and endpoints watchers are same. I think it's better to make another abstract layer and let each mode only focus on core logic.

@feiskyer I agree...

dcbw commented:
@thockin @madhanrm there are still outstanding review comments, so IMO @thockin shouldn't have lgtm-ed it without explaining why those review comments either don't apply or should be punted...  Tim, thoughts on that?

m1093782566 commented:
> There are much code repetition across different proxy modes, e.g. the services and endpoints watchers are same. I think it's better to make another abstract layer and let each mode only focus on core logic.

+1

madhanrm commented:
For Cross compilation, there is an issue with the case sensitive collision of sirupsen/logrus
https://github.com/sirupsen/logrus/issues/451
Am not sure how to fix this. It looks like a separate PR is required to fix the case collision, If I vendor in the changes along with this PR.

madhanrm commented:
/retest

cmluciano commented on cmd/kube-proxy/app/server.go:
```diff
@@ -60,37 +58,33 @@ import (
+	proxyModeUserspace   = "userspace"
+	proxyModeIPTables    = "iptables"
+	proxyModeIPVS        = "ipvs"
+	proxyModeKernelspace = "kernelspace"
```
Not sure if this has been mentioned but kernelspace does not translate to windowsmode to me.

madhanrm replied:
This is used in analogous to userspace. Doesn't necessarily map to windows mode. Can be reused in Linux as well.

bowei replied:
"kernelspace" doesn't align with the other modes, as they all name specific technologies used (e.g. IPVS, IPTables, ...)

cmluciano commented on cmd/kube-proxy/app/server.go:
```diff
@@ -776,6 +548,10 @@ func (s *ProxyServer) Run() error {
+func (s *ProxyServer) birthCry() {
+	s.Recorder.Eventf(s.NodeRef, api.EventTypeNormal, "Starting", "Starting kube-proxy.")
+}
+
```
Is this a golint recommendation?

madhanrm replied:
This is existing code. Am not sure I understand your question

madhanrm commented:
/rerun

madhanrm commented:
/retest

madhanrm commented:
/test pull-kubernetes-e2e-gce-bazel

fejta commented:
/retest
ref https://github.com/kubernetes/test-infra/issues/4568

madhanrm commented:
/test pull-kubernetes-e2e-gce-bazel

madhanrm commented:
/rerun

madhanrm commented:
/rerun

m1093782566 commented:
/retest

madhanrm commented:
/retest

madhanrm commented:
Fixed the cross compile issues. Waiting on the pending test to complete and a lgtm.

thockin commented:
Assign back to me once it is LGTM'ed and ready for approval.

dims commented:
/retest

madhanrm commented:
/retest

dims commented:
/test all

dims commented:
@madhanrm Can you please check the test failures and get the tests go green? i am afraid this will miss the 1.8 boat otherwise

dineshgovindasamy commented:
/retest

dineshgovindasamy commented:
/retest

thockin commented:
Is this ready?

dims commented:
the kubeadm-gce is stuck in "job triggered", trying to get that to run

/test all

dineshgovindasamy commented:
@dims @thockin i have tried multiple retest and Kubeadm-gce is not at all getting triggered. I see that job is failing across multiple PR's. How do we proceed?

dims commented:
@fejta @spiffxp any tips here?

krzyzacy commented:
https://github.com/kubernetes/test-infra/pull/4649

but do you care about the kubeadm job on your PR? that's a non-blocking job.

krzyzacy commented:
If it's ready, ask for a lgtm label and this can be picked up by sq/get merged regardless of kubeadm presubmit.

dims commented:
Looks like @thockin applied lgtm at least 2 times... and it was rebased away. let me try doing that

/assign

dims commented:
/lgtm

BenTheElder commented:
It looks like the kubeadm job failed and just never posted the status. /cc @luxas.
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/51064/pull-kubernetes-e2e-kubeadm-gce/7151/

dims commented:
/test pull-kubernetes-unit
